### ISSUE TITLE
Warn when f-string with non-local variable used? Warn when f-string converting function to str is used?

### ISSUE THREAD

**Issue by mpasternak** [gforcada/flake8-builtins#105]
Hello, 

this is the third time I am submitting this bug, hopefully this is the right place, @jack-mcivor directed me here from https://github.com/PyCQA/pyflakes/issues/787 , before I submitted it to flake8 repo...

I am not sure if this belongs to this plugin and I'm not sure how to bite this exactly, but I have feeling somewhere in the back of my head, that this should be made visible to the end-user, aka programmer. 

If this does not belong to this plugin, please help me and show me the direction I should take it to. 

I recently made a somewhat funny bug in my codebase. 

It looked like this: 

`SOME_CONST = f"http://some.host/{id}"
`

The bug-introducing character was that little "f" before the starting quote. 

I meant to declare a string, that gets `.format`'ted some time later (like first I defined the URL with the place for the parameters, then I wanted to do `SOME_CONST.format({"id": "some_id"})`)

Instead, I declared a constant that looked like `http://some.host/<builtin function id>`. To add to the injury, `.format()` on that constant worked without problems at all, obviously. 

First I wonder what would I like to do with that. We already have a warning for when there are f-string format characters and the string is missing the starting "f". 

This would be the opposite, but only for some symbols. 

On the other hand, there is very little chance somebody would want to have a string-description of a function embedded in a string constant. 

So, warning about a variable found in f-string not declared in the local scope could be okay, but I am not sure if this is even possible. That `SOME_CONST` was declared at the top level of the module, all the imported things and symbols were available. Of course that `id()` was in the builtins, not in the module scope. 

Maybe it would be sensible to warn the user about converting function-type object to a string in a string constant. I think that this is not the thing people like to do on the daily basis. 

I would like to know your opinion on this problem first.

Any opinions?

**Comment by gforcada** [comment]
Thanks for the report! 👍🏾 and specially if it's the third time you are reporting it 😓 

So the logic you search to be reported is something like:

_Warn me if an f-string is then used with `.format`_ ?

Although I see perfectly valid examples of doing that as well 🤷🏾 

I see that there is https://github.com/MichaelKim0407/flake8-use-fstring which might be more fitting to your use case, or maybe this other one? https://pypi.org/project/flake8-string-format/ ?

Bottom line is that, the heuristics to decide if a string should, or should not have the `f` prefix can be very blurry, and all use cases at some point might be valid for ones, or introduce bugs on others ⚖️ 

As for this specific flake8 plugin, well, we could indeed, warn if a placeholder within an f-string is a builtin, that might be worth reporting, or? 🤔

**Comment by mpasternak** [comment]
> So the logic you search to be reported is something like:

> Warn me if an f-string is then used with *.format* ?

No.  

This is actually my use-case and I do agree with you, this is perfectly valid. In my code I declared strings with curly brackets and I wanted to format them later. Like I wrote earlier, I wanted `SOME_URL = "https://foobar/{id}"` and then I used `SOME_URL.format({"id": "so-and-so"})`. This is valid. 

What was not valid and what I perceive as dangerous was, that at some point me or my IDE added an 'f' before quotes. PyCharm does that sometimes, when it thinks that I am writing an f-string, and I believe that it could think that I am creating an f-string so maybe it added this. Perhaps it is worth adding to PyCharm bug tracker as well. 

And so I ended up with `SOME_URL = *f*"https://foobar/{id}"`. And `id` is a built-in function of course. So it ended up being `SOME_URL = "https://foobar/<builtin function id>"`. 

In my opinion what would be most fitting for this case is conversion of a built-in function to a string form. My code actually executed `str(id)` and the more I think about it, the more I think that there is actually **no valid use-case for converting function to a string**, except for tracebacks, sentry.io, interactive console, interactive debugging. So... in other words, calling `str(id)` in a production app will always (in my opinion) signify that this production app stops being in a production state and goes more to an interactive state (or "asynchronously interactive" as in send an error e-mail, notify sentry.io, etc). 

So by extrapolation I don't see `str(built-in-function)` happening in the code at the import time any good. 

I also saw the other flake8 plugin modules, thank you very much for pointing them out. 

Somehow I think that as this bug has to do with built-in function, well... :-)

Do you think it could be possible to prepare a patch especially for your plugin? I could put in some work to have this implemented... as soon as you OK this. Thanks!

**Comment by gforcada** [comment]
Uh, sorry, it took me some time to read your last comment...

Now, I think 🤞🏾 , I do understand a bit better your concern.

That's a difficult one though 🤔 I mean, checking if a string, or an f-string has a builtin within its parameters...

I'm not 100% sure, but until Python 3.12 there was not defined grammar for f-strings, so I'm not sure if one can get an `ast` out of them with any previous version.

IMHO though, this really does not belong to this plugin: indeed, is about builtins, but it is much more about how f-strings, or strings in general behave.

I would be happy to help on creating a new `flake8` plugin that does that, please copy the infrastructure/files I created for this plugin for yours 😉

I think it would be best to have a separate one, or look for other `flake8` plugins that deal with f-string formattings 👍🏾

**Comment by mpasternak** [comment]
Thank you for the opinion. 

I gave it some thought also.  I think that calling \_\_str__ for any builtin during runtime should be treated as suspicious. I can imagine some use-cases for it, but I don't see it being anything of value during runtime on a production site. I will consider creating a plugin for it, yours will surely be used as a base of it, thanks.
